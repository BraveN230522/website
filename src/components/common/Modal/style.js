import React from 'react';
import styled from '@emotion/styled';
import { keyframes } from '@emotion/core';

import Icon from '@components/common/Icon';
import mobileLogo from '@svg/logo-symbol-only.svg';

const revealAnimation = keyframes`
  0%, 3% {
    opacity: 0;
  }
 
  4%, 30% {
    opacity: 0.3;
  }

  60% {
    opacity: 0.8;
  }

  90% {
    opacity: 1;
  }
`;

const ModalWrapper = styled.div`
  position: fixed;
  z-index: 999;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  background-color: ${({ theme }) => theme.colors.overlay};
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  ${({ theme }) => theme.media.phone`
    bottom: initial;
    min-height: 100vh;
  `}
`;

const ContentWrapper = styled.div`
  overflow-y: auto; /* Enable scroll if needed (for the content only ) */
  -webkit-overflow-scrolling: touch;
  position: relative;
  z-index: 99;
  width: 800px;
  min-height: 472px;

  background: ${({ theme }) => theme.colors.white};
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 16px;
  box-shadow: 0px 1px 16px rgba(0, 0, 0, 0.2);

  ${({ theme }) =>
    theme.generateBreakpoint(
      768,
      `
    flex-direction: column;
    width: 100%;
    height: 100%;
    border-radius: 0;
    box-shadow: none;
    max-height: unset;
  `
    )}

    ${({ theme }) => theme.media.phone`
      height: 100vh;
    `}

    animation: ${revealAnimation} 0.4s ease;
  `;

const CloseButtonWrapper = styled.a`
    position: absolute;
    bottom: 21px;
    left: 17px;
    
    :hover,
    :focus {
      cursor: pointer;
`;

const CloseButton = ({ onClick }) => (
  <CloseButtonWrapper onClick={onClick} data-testid="modal-close-button">
    <Icon name="close" color="#FFFFFF" />
  </CloseButtonWrapper>
);

const ModalHeaderWrapper = styled.div`
  position: relative;
  height: ${({ theme }) => theme.mobileHeaderHeight};
  width: 100vw;
  display: none;

  background-color: ${({ theme }) => theme.colors.swiamBlue};

  ${({ theme }) =>
    theme.generateBreakpoint(
      768,
      `
      display: flex;
      justify-content: center;
      align-items: flex-end;
      padding-bottom: 12px;
  `
    )}
`;

const ModalHeader = ({ onClick }) => (
  <ModalHeaderWrapper data-testid="modal-header-mobile">
    <CloseButton onClick={onClick} />
    <img src={mobileLogo} alt="Sports where I am" />
  </ModalHeaderWrapper>
);

export { ModalWrapper, ContentWrapper, ModalHeader };
