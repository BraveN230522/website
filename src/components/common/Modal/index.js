import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import StopBodyScroll from '@components/common/StopBodyScroll';

import { ModalWrapper, ModalHeader, ContentWrapper } from './style';

const Modal = ({ children, isInitiallyOpened, onOpening, onClosing }) => {
  const [isOpened, setOpeningStatus] = useState(isInitiallyOpened);

  useEffect(() => {
    setOpeningStatus(isInitiallyOpened);
  }, [isInitiallyOpened]);

  useEffect(() => {
    if (isOpened) {
      onOpening();
      setOpeningStatus(true);
    }
  }, [isOpened]);

  const handleClosingClick = useCallback(() => {
    onClosing();
    setOpeningStatus(false);
  }, []);

  const handleDismiss = useCallback(e => {
    if (e.target.id === 'dismiss-modal') {
      handleClosingClick();
    }
  }, []);

  if (!isOpened) {
    return null;
  }

  return (
    <>
      <StopBodyScroll />
      <ModalWrapper
        id="dismiss-modal"
        onClick={handleDismiss}
        role="dialog"
        aria-modal={isOpened}
        data-testid="modal-component"
      >
        <ContentWrapper>
          <ModalHeader onClick={handleClosingClick} />

          {Boolean(children) && children(setOpeningStatus)}
        </ContentWrapper>
      </ModalWrapper>
    </>
  );
};

Modal.defaultProps = {
  isInitiallyOpened: false,
  onOpening: () => {},
  onClosing: () => {},
};

Modal.propTypes = {
  isInitiallyOpened: PropTypes.bool,
  onOpening: PropTypes.func,
  onClosing: PropTypes.func,
};

export default Modal;
