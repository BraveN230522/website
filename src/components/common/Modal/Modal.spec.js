import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import withGlobals from '@swiamTestUtils/withGlobals';
import { matchers } from 'jest-emotion';

import Modal from './index';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);

afterEach(() => {
  cleanup();
});

const testId = 'modal-component';
const childrenTestId = 'modal-component-children';

const MockedModalChildren = () => (
  <div data-testid={childrenTestId} style={{ width: '100%', height: '100%' }}>
    This is a mocked children content
  </div>
);

describe('<Modal />', () => {
  it('Renders the Modal unchanged', () => {
    const { asFragment, queryByTestId } = render(<Modal />, {
      wrapper: withGlobals,
    });
    expect(asFragment()).toMatchSnapshot(); // Save the snapshot of a null component
    expect(queryByTestId(testId)).toBeNull();
  });

  it("Hide modal when 'isInitiallyOpened' is false", () => {
    const { queryByTestId } = render(<Modal isInitiallyOpened={false} />, {
      wrapper: withGlobals,
    });
    expect(queryByTestId(testId)).toBeNull();
  });

  it("Show modal when 'isInitiallyOpened' is true", () => {
    const { queryByTestId, getByTestId, asFragment } = render(<Modal isInitiallyOpened />, {
      wrapper: withGlobals,
    });
    expect(queryByTestId(testId)).toBeTruthy();
    expect(asFragment()).toMatchSnapshot(); // Save the snapshot of a fully rendered component
    expect(getByTestId(testId).getAttribute('aria-modal')).toBe('true');
  });

  it("Call 'onOpening' once if callback is provided", () => {
    const onOpening = jest.fn();

    render(<Modal isInitiallyOpened onOpening={onOpening} />, {
      wrapper: withGlobals,
    });
    expect(onOpening).toHaveBeenCalledTimes(1);
  });

  it('Keep Modal open when clicked into children content', () => {
    let open = true;
    const onClosing = jest.fn(() => {
      open = false;
    });
    const { getByTestId } = render(
      <Modal isInitiallyOpened onClosing={onClosing}>
        {() => <MockedModalChildren />}
      </Modal>,
      {
        wrapper: withGlobals,
      }
    );

    fireEvent.click(getByTestId(childrenTestId));

    expect(open).toBe(true);
    expect(getByTestId(testId).getAttribute('aria-modal')).toBe('true');
  });

  it('Dismiss desktop Modal when clicked outside children content', () => {
    let open = true;
    const onClosing = jest.fn(() => {
      open = false;
    });
    const { getByTestId, queryByTestId } = render(
      <Modal isInitiallyOpened onClosing={onClosing}>
        {() => <MockedModalChildren />}
      </Modal>,
      {
        wrapper: withGlobals,
      }
    );

    fireEvent.click(getByTestId(testId));

    expect(open).toBe(false);
    expect(queryByTestId(testId)).toBeNull();
  });

  it('Dismiss mobile Modal when clicked on the close button', () => {
    let open = true;
    const onClosing = jest.fn(() => {
      open = false;
    });
    const { getByTestId, queryByTestId } = render(
      <Modal isInitiallyOpened onClosing={onClosing}>
        {() => <MockedModalChildren />}
      </Modal>,
      {
        wrapper: withGlobals,
      }
    );

    fireEvent.click(getByTestId('modal-close-button'));

    expect(open).toBe(false);
    expect(queryByTestId(testId)).toBeNull();
  });

  it('Show modal header only on mobile', () => {
    const { getByTestId } = render(<Modal isInitiallyOpened>{() => <MockedModalChildren />}</Modal>, {
      wrapper: withGlobals,
    });

    const elm = getByTestId('modal-header-mobile');

    expect(elm).toHaveStyleRule('display', 'none');
    expect(elm).toHaveStyleRule('display', 'flex', {
      media: `(max-width: 48em)`,
    });
  });
});
