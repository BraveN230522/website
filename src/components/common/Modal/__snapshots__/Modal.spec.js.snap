// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> Renders the Modal unchanged 1`] = `<DocumentFragment />`;

exports[`<Modal /> Show modal when 'isInitiallyOpened' is true 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  0%, 3% {
    opacity: 0;
  }

  4%, 30% {
    opacity: 0.3;
  }

  60% {
    opacity: 0.8;
  }

  90% {
    opacity: 1;
  }
}

.emotion-6 {
  position: fixed;
  z-index: 999;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.5);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  overflow-y: auto;
  -webkit-overflow-scrolling: touch;
  position: relative;
  z-index: 99;
  width: 800px;
  min-height: 472px;
  background: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 16px;
  box-shadow: 0px 1px 16px rgba(0,0,0,0.2);
  -webkit-animation: animation-0 0.4s ease;
  animation: animation-0 0.4s ease;
}

@media (max-width:48em) {
  .emotion-4 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    width: 100%;
    height: 100%;
    border-radius: 0;
    box-shadow: none;
    max-height: unset;
  }
}

.emotion-2 {
  position: relative;
  height: 93px;
  width: 100vw;
  display: none;
  background-color: #2891ff;
}

@media (max-width:48em) {
  .emotion-2 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-align-items: flex-end;
    -webkit-box-align: flex-end;
    -ms-flex-align: flex-end;
    align-items: flex-end;
    padding-bottom: 12px;
  }
}

.emotion-0 {
  position: absolute;
  bottom: 21px;
  left: 17px;
}

.emotion-0:hover,
.emotion-0:focus {
  cursor: pointer;
}

<div
    aria-modal="true"
    class="emotion-6 emotion-7"
    data-testid="modal-component"
    id="dismiss-modal"
    role="dialog"
  >
    <div
      class="emotion-4 emotion-5"
    >
      <div
        class="emotion-2 emotion-3"
        data-testid="modal-header-mobile"
      >
        <a
          class="emotion-0 emotion-1"
          data-testid="modal-close-button"
        >
          <svg
            color="#FFFFFF"
            height="24"
            width="24"
          >
            <defs>
              <path
                d="M.293.293c.39-.39 1.024-.39 1.414 0L7 5.585 12.293.293c.36-.36.928-.388 1.32-.083l.094.083c.39.39.39 1.024 0 1.414L8.415 7l5.292 5.293c.36.36.388.928.083 1.32l-.083.094c-.39.39-1.024.39-1.414 0L7 8.415l-5.293 5.292c-.36.36-.928.388-1.32.083l-.094-.083c-.39-.39-.39-1.024 0-1.414L5.585 7 .293 1.707C-.067 1.347-.095.78.21.387z"
                id="prefix__a"
              />
            </defs>
            <g
              fill="none"
              fill-rule="evenodd"
              transform="translate(5 5)"
            >
              <mask
                fill="#FFFFFF"
                id="prefix__b"
              >
                <use
                  xlink:href="#prefix__a"
                />
              </mask>
              <use
                fill="#FFFFFF"
                fill-rule="nonzero"
                xlink:href="#prefix__a"
              />
              <g
                fill="#FFFFFF"
                mask="url(#prefix__b)"
              >
                <path
                  d="M0 0H24V24H0z"
                  transform="translate(-5 -5)"
                />
              </g>
            </g>
          </svg>
        </a>
        <img
          alt="Sports where I am"
          src="mocked-file"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
